readparameters.h

```diff
@@ -22,57 +22,169 @@
+   ~Readparameters();
+
+   /** Add a new input parameter.
+    * Note that parse must be called in order for the input file(s) to be re-read.
+    * Only called by the root process.
+    * @param name The name of the parameter, as given in the input file(s).
+    * @param desc Description for the parameter.
+    * @param defValue Default value for variable.
+    */
+   static void add(const std::string& name, const std::string& desc, const std::string& defValue) {
+      int rank;
+      MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+      if (rank == MASTER_RANK) {
+         options[name] = "";
+         isOptionParsed[name] = false;
+         descriptions->add_options()(name.c_str(), PO::value<std::string>(&(options[name]))->default_value(defValue),
+                                     desc.c_str());
+      }
+   }
+
+   template <typename T> static void add(const std::string& name, const std::string& desc, const T& defValue) {
+      int rank;
+      MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+      if (rank == MASTER_RANK) {
+         std::stringstream ss;
+
+         if constexpr (std::is_floating_point_v<T>) {
+            ss << std::setprecision(std::numeric_limits<double>::digits10 + 1) << defValue;
+         } else {
+            ss << defValue;
+         }
+         options[name] = "";
+         isOptionParsed[name] = false;
+         descriptions->add_options()(name.c_str(), PO::value<std::string>(&(options[name]))->default_value(ss.str()),
+                                     desc.c_str());
+      }
+   }
+
+   /** Get the value of the given parameter added with add().
+    * This may be called after having called Parse, and it may be called by any process, in any order.
+    * @param name The name of the parameter.
+    * @param value A variable where the value of the parameter is written.
+    * @return If true, the given parameter was found and its value was written to value.
+    */
+   static void get(const std::string& name, std::string& value) {
+      if (options.find(name) != options.end()) { // check if it exists
+         value = options[name];
+      } else {
+         int rank;
+         MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+         if (rank == MASTER_RANK) {
+            std::cerr << __FILE__ << ":" << __LINE__ << name + " not found in the configuration file!" << std::endl;
+            MPI_Abort(MPI_COMM_WORLD, 1);
+         }
+      }
+   }
+
+   static void get(const std::string& name, std::vector<std::string>& value) {
+      if (vectorOptions.find(name) != vectorOptions.end()) { // check if it exists
+         value = vectorOptions[name];
+      } else {
+         int rank;
+         MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+         if (rank == MASTER_RANK) {
+            std::cerr << __FILE__ << ":" << __LINE__ << name + " not found in the configuration file!" << std::endl;
+            MPI_Abort(MPI_COMM_WORLD, 1);
+         }
+      }
+   }
+
+   template <typename T> static void get(const std::string& name, T& value) {
```

Comment:
Hmm, how does the return type (void) play with our existing use pattern here? In all old code, ```RP::get()``` is called like
```
if(!RP::get("name",target)) {
   std::cerr << "Some error message" << std::endl;
}
```
So if this becomes void, that line will throw an error.

Since you changed all calls to this in all projects anyway (awesome work! I think it makes the code a lot better), but I'm wondering how this'll affect any as-yet-unmerged pull requests. Three options:
 - If we leave this to be a bool, they will probably simply work. But their code will probably never be cleaned up, because why would it?
 - If we leave it as void, they will throw errors, and whoever works on them gets to fix their code once they merge.
 - The ideal case would be to set it to bool and make the compiler emit a warning if it is actually used. Something like the opposite of [the nodiscard attribute](https://en.cppreference.com/w/cpp/language/attributes/nodiscard) or the opposite of [GCC's warn_unused_result](https://gcc.gnu.org/onlinedocs/gcc-4.1.1/gcc/Function-Attributes.html) attribute. Except these don't seem to exist. Ho-hum.

Reply:
I notice that in some of the projects, the `get` function in current dev is called as
```c++
RP::get("name", target)
```
which does not follow the design pattern but it works as long as no error in reading occurs.

I don't know which is a proper way to handle this backward compatibility like problem. Leave for you to decide.